### ISSUE TITLE
[core]  IndexOutOfBoundsException in RegionMakerVisitor

### ISSUE THREAD

**Issue by huku-** [skylot/jadx#799]
Hello,

Since this is the first time I'm opening an issue in jadx, let me take some time to say thank you for keeping jadx open-source and freely available to us. I guess you have probably heard this prose a million times, but I really mean it.

So, I got back to using jadx again after a small break and I thought it would be nice to start looking into the problems that cause the "Method not decompiled" exceptions in the output. Unfortunately, due to the nature of my work, I cannot disclose the name of the APK I'm analyzing, but I'll try to be as precise as possible.

Several of the decompilation errors that I get look like this:

```
JADX ERROR: IndexOutOfBoundsException in pass: RegionMakerVisitor
java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
    at java.util.ArrayList.rangeCheck(ArrayList.java:657)
    at java.util.ArrayList.get(ArrayList.java:433)
    at jadx.core.dex.nodes.InsnNode.getArg(InsnNode.java:101)
    at jadx.core.dex.visitors.regions.RegionMaker.traverseMonitorExits(RegionMaker.java:611)
    at jadx.core.dex.visitors.regions.RegionMaker.traverseMonitorExits(RegionMaker.java:619)
    at jadx.core.dex.visitors.regions.RegionMaker.traverseMonitorExits(RegionMaker.java:619)
    at jadx.core.dex.visitors.regions.RegionMaker.traverseMonitorExits(RegionMaker.java:619)
    at jadx.core.dex.visitors.regions.RegionMaker.traverseMonitorExits(RegionMaker.java:619)
    at jadx.core.dex.visitors.regions.RegionMaker.traverseMonitorExits(RegionMaker.java:619)
    at jadx.core.dex.visitors.regions.RegionMaker.traverseMonitorExits(RegionMaker.java:619)
    at jadx.core.dex.visitors.regions.RegionMaker.traverseMonitorExits(RegionMaker.java:619)
    at jadx.core.dex.visitors.regions.RegionMaker.traverseMonitorExits(RegionMaker.java:619)
    at jadx.core.dex.visitors.regions.RegionMaker.traverseMonitorExits(RegionMaker.java:619)
    at jadx.core.dex.visitors.regions.RegionMaker.traverseMonitorExits(RegionMaker.java:619)
    at jadx.core.dex.visitors.regions.RegionMaker.traverseMonitorExits(RegionMaker.java:619)
    at jadx.core.dex.visitors.regions.RegionMaker.processMonitorEnter(RegionMaker.java:561)
    at jadx.core.dex.visitors.regions.RegionMaker.traverse(RegionMaker.java:133)
    at jadx.core.dex.visitors.regions.RegionMaker.makeRegion(RegionMaker.java:86)
    at jadx.core.dex.visitors.regions.RegionMaker.processIf(RegionMaker.java:693)
    at jadx.core.dex.visitors.regions.RegionMaker.traverse(RegionMaker.java:123)
    at jadx.core.dex.visitors.regions.RegionMaker.makeRegion(RegionMaker.java:86)
    at jadx.core.dex.visitors.regions.RegionMaker.processIf(RegionMaker.java:693)
    at jadx.core.dex.visitors.regions.RegionMaker.traverse(RegionMaker.java:123)
    at jadx.core.dex.visitors.regions.RegionMaker.makeRegion(RegionMaker.java:86)
    at jadx.core.dex.visitors.regions.RegionMakerVisitor.visit(RegionMakerVisitor.java:49)
```

The problem seems to be in `traverseMonitorExits()` in **RegionMaker.java**:

```java
for (InsnNode insn : block.getInstructions()) {
    if (insn.getType() == InsnType.MONITOR_EXIT
            && insn.getArg(0).equals(arg)) {
        exits.add(block);
        region.getExitInsns().add(insn);
        return;
    }
}
```

The snippet above assumes that there's at least one argument each `MONITOR_EXIT` instruction, which is usually fine. However, sometimes this might not be the case, as mentioned in a comment in `processMonitorEnter()`.

```java
// remove arg from MONITOR_EXIT to allow inline in MONITOR_ENTER
exitInsn.removeArg(0);
exitInsn.add(AFlag.DONT_GENERATE);
```

To solve this problem I just added a check for the number of arguments of `MONITOR_EXIT` instructions in `traverseMonitorExits()`:

```java
for (InsnNode insn : block.getInstructions()) {
    if (insn.getType() == InsnType.MONITOR_EXIT
            && insn.getArgsCount() > 0
            && insn.getArg(0).equals(arg)) {
        exits.add(block);
        region.getExitInsns().add(insn);
        return;
    }
}
```

I guess you could also check for `DONT_GENERATE`, but haven't tried that.

Anyway, I'm not creating a pull-request as I'm not really sure if this is the most elegant way of fixing this bug. If there's a better way to do it, I'll be happy to know. In the mean time, I'll be studying the source code to determine if there's a deeper root cause that results in this exception.

**Comment by skylot** [comment]
@huku- thanks for report and analysis!
Additional check is good, but for me it is unclear why this can happen because arg removing is *after* `traverseMonitorExits()` call. And as soon as I don't know how to reproduce this it is hard to tell exacly what happens here.
@huku- if you can please provide some java code sample or maybe an integration test like [this](https://github.com/skylot/jadx/blob/84cb6b9569d998548b333e0d388c14d0a75aeb6e/jadx-core/src/test/java/jadx/tests/integration/synchronize/TestSynchronized.java#L13) so I can reproduce this issue.

**Comment by huku-** [comment]
Oki doki,

I managed to narrow this down a bit. Here's the final smali test-case:

```
.class public final Ltest/TestReceiver;
.super Landroid/content/BroadcastReceiver;
.source "TestReceiver.java"

.method public final onReceive(Landroid/content/Context;Landroid/content/Intent;)V
    .locals 2
    const/4 v0, 0
    const/4 v1, 0
    monitor-enter v0
    monitor-enter v1
    monitor-exit v1
    monitor-exit v0
    return-void
.end method
```

I use the following commands to reproduce:

```
java -jar smali.jar assemble -a 27 -o test.dex TestReceiver.smali
jadx test.dex
```

The exception goes away if I remove the nested monitor exit, so, I guess we found the root cause? I haven't had a look at the source code yet as it took me a while to minimize this one.

**Comment by skylot** [comment]
@huku- thanks for the test case.
Java code:
```java
public void test() {
	Object obj1 = null;
	Object obj2 = null;
	synchronized (obj1) {
		synchronized (obj2) {
		}
	}
}
```
This code is weird because you can't synchronized on null values.
And as soon as jadx applies constant inlining before any code transformations all `monitor-enter`/`monitor-exit` pairs will have a same null argument and it is impossible to match `monitor-enter` with corresponding `monitor-exit` and that leads to an exception. 

Anyway, I made a fix which forbids constant inlining in synchronized statement and jadx starts producing correct code.

**Comment by huku-** [comment]
Yes, the code is indeed weird because I removed all variables, values and code that was not relevant to the bug. `v0` and `v1` were originally instances of some classes, but I replaced their values with 0.

> Anyway, I made a fix which forbids constant inlining in synchronized statement and jadx starts producing correct code.

Maybe this was the case with the original code as well (i.e. constant values ending up in monitor instructions). I will run more tests and let you know.
